File: vsg_core/subtitles/frame_utils/timing.py
# vsg_core/subtitles/frame_utils/timing.py
"""
Frame/time conversion functions for subtitle synchronization.

Contains:
- CFR timing modes (floor, middle, aegisub)
- VFR (VideoTimestamps-based) timing
- VFR cache management
"""

from __future__ import annotations

import gc
import math
import threading
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from vsg_core.models import AppSettings

# ============================================================================
# MODE 0: FRAME START (For Correlation-Frame-Snap - STABLE & DETERMINISTIC)
# ============================================================================


def time_to_frame_floor(time_ms: float, fps: float) -> int:
    """
    MODE: Frame START (stable, deterministic).

    Convert timestamp to frame number using FLOOR with epsilon protection.
    This gives the frame that is currently displaying at the given time.

    This is the preferred method for sync math because:
    - Deterministic (no rounding ambiguity at boundaries)
    - Stable under floating point drift
    - Maps to actual frame boundaries (frame N starts at N * frame_duration)

    Args:
        time_ms: Timestamp in milliseconds
        fps: Frame rate (e.g., 23.976)

    Returns:
        Frame number (which frame is displaying at this time)

    Examples at 23.976 fps (frame_duration = 41.708ms):
        time_to_frame_floor(0.0, 23.976) -> 0
        time_to_frame_floor(41.707, 23.976) -> 0 (still in frame 0)
        time_to_frame_floor(41.708, 23.976) -> 1 (frame 1 starts)
        time_to_frame_floor(1000.999, 23.976) -> 23 (FP drift protected)
        time_to_frame_floor(1001.0, 23.976) -> 24
    """
    frame_duration_ms = 1000.0 / fps
    # Add small epsilon to protect against FP errors where time_ms is slightly under frame boundary
    epsilon = 1e-6
    return int((time_ms + epsilon) / frame_duration_ms)


def frame_to_time_floor(frame_num: int, fps: float) -> float:
    """
    MODE: Frame START (stable, deterministic).

    Convert frame number to its START timestamp (exact, no rounding).

    This is the preferred method for sync math because:
    - Frame N starts at exactly N * frame_duration
    - No rounding (exact calculation)
    - Guarantees frame-aligned timing

    Args:
        frame_num: Frame number
        fps: Frame rate (e.g., 23.976)

    Returns:
        Timestamp in milliseconds (frame START time, as float for precision)

    Examples at 23.976 fps (frame_duration = 41.708ms):
        frame_to_time_floor(0, 23.976) -> 0.0
        frame_to_time_floor(1, 23.976) -> 41.708
        frame_to_time_floor(24, 23.976) -> 1001.0
        frame_to_time_floor(100, 23.976) -> 4170.8
    """
    frame_duration_ms = 1000.0 / fps
    return frame_num * frame_duration_ms


# ============================================================================
# MODE 1: MIDDLE OF FRAME (Current Implementation)
# ============================================================================


def time_to_frame_middle(time_ms: float, fps: float) -> int:
    """
    MODE: Middle of frame window.

    Convert timestamp to frame number, accounting for +0.5 offset.

    Args:
        time_ms: Timestamp in milliseconds
        fps: Frame rate (e.g., 23.976)

    Returns:
        Frame number
    """
    frame_duration_ms = 1000.0 / fps
    return round(time_ms / frame_duration_ms - 0.5)


def frame_to_time_middle(frame_num: int, fps: float) -> int:
    """
    MODE: Middle of frame window.

    Targets the middle of the frame's display window with +0.5 offset.

    Example at 23.976 fps:
    - Frame 24 displays from 1001.001ms to 1042.709ms
    - Calculation: 24.5 x 41.708 = 1022ms
    - After centisecond rounding: 1020ms (safely in frame 24)

    Args:
        frame_num: Frame number
        fps: Frame rate (e.g., 23.976)

    Returns:
        Timestamp in milliseconds
    """
    frame_duration_ms = 1000.0 / fps
    return int(round((frame_num + 0.5) * frame_duration_ms))


# ============================================================================
# MODE 2: AEGISUB-STYLE (Ceil to Centisecond)
# ============================================================================


def time_to_frame_aegisub(time_ms: float, fps: float) -> int:
    """
    MODE: Aegisub-style timing.

    Convert timestamp to frame using floor division (which frame is currently displaying).

    Args:
        time_ms: Timestamp in milliseconds
        fps: Frame rate

    Returns:
        Frame number
    """
    frame_duration_ms = 1000.0 / fps
    return int(time_ms / frame_duration_ms)


def frame_to_time_aegisub(frame_num: int, fps: float) -> int:
    """
    MODE: Aegisub-style timing.

    Matches Aegisub's algorithm: Calculate exact frame start, then round UP
    to the next centisecond to ensure timestamp falls within the frame.

    Example at 23.976 fps:
    - Frame 24 starts at 1001.001ms
    - Exact calculation: 24 x 41.708 = 1001.001ms
    - Round UP to next centisecond: ceil(1001.001 / 10) x 10 = 1010ms
    - Result: 1010ms (safely in frame 24: 1001-1043ms)

    Args:
        frame_num: Frame number
        fps: Frame rate

    Returns:
        Timestamp in milliseconds
    """
    frame_duration_ms = 1000.0 / fps
    exact_time_ms = frame_num * frame_duration_ms

    # Round UP to next centisecond (ASS format precision)
    # This ensures the timestamp is guaranteed to fall within the frame
    centiseconds = math.ceil(exact_time_ms / 10)
    return centiseconds * 10


# ============================================================================
# MODE 3: VFR (VideoTimestamps-based)
# ============================================================================

# Cache for VideoTimestamps instances to avoid re-parsing video
# Thread-safe: accessed from ThreadPoolExecutor workers
_vfr_cache = {}
_vfr_cache_lock = threading.Lock()


def clear_vfr_cache():
    """
    Clear the VFR cache to release VideoTimestamps instances.

    This should be called on application shutdown or when clearing resources
    to prevent nanobind reference leaks.
    """
    global _vfr_cache
    with _vfr_cache_lock:
        _vfr_cache.clear()
    gc.collect()  # Force garbage collection to release nanobind objects


def get_vfr_timestamps(
    video_path: str, fps: float, runner, settings: AppSettings | None = None
):
    """
    Get appropriate timestamp handler based on video type.

    For CFR videos: Uses lightweight FPSTimestamps (just calculations)
    For VFR videos: Uses VideoTimestamps (analyzes actual video)

    Args:
        video_path: Path to video file
        fps: Frame rate
        runner: CommandRunner for logging
        settings: AppSettings with videotimestamps_rounding setting
    """
    try:
        from fractions import Fraction

        from video_timestamps import (
            FPSTimestamps,
            RoundingMethod,
        )

        # Get rounding method from settings (default: ROUND)
        rounding_str = (
            settings.videotimestamps_rounding.upper()
            if settings is not None
            else "ROUND"
        )

        if rounding_str == "FLOOR":
            rounding_method = RoundingMethod.FLOOR
        else:  # 'ROUND' or default
            rounding_method = RoundingMethod.ROUND

        # Create cache key that includes rounding method
        cache_key = f"{video_path}_{rounding_str}"

        # Thread-safe cache access
        with _vfr_cache_lock:
            if cache_key in _vfr_cache:
                return _vfr_cache[cache_key]

        # Try to detect if video is VFR by checking if it's a real video file
        # For now, use FPSTimestamps (lightweight) for CFR videos
        # This just does math, doesn't analyze the video file

        # Convert FPS to exact fraction for NTSC drop-frame rates
        # NTSC standards use fractional rates (N*1000/1001) to avoid color/audio drift
        if abs(fps - 23.976) < 0.001:
            fps_frac = Fraction(
                24000, 1001
            )  # 23.976fps - NTSC film (24fps slowed down)
        elif abs(fps - 29.97) < 0.01:
            fps_frac = Fraction(
                30000, 1001
            )  # 29.97fps - NTSC video (30fps slowed down)
        elif abs(fps - 59.94) < 0.01:
            fps_frac = Fraction(
                60000, 1001
            )  # 59.94fps - NTSC high fps (60fps slowed down)
        else:
            # Use decimal FPS as fraction for non-NTSC rates (PAL, web video, etc.)
            fps_frac = Fraction(int(fps * 1000), 1000).limit_denominator(10000)

        # Use FPSTimestamps for CFR (constant framerate) - lightweight!
        time_scale = Fraction(1000)  # milliseconds
        vts = FPSTimestamps(rounding_method, time_scale, fps_frac)

        runner._log_message(
            f"[VideoTimestamps] Using FPSTimestamps for CFR video at {fps:.3f} fps"
        )
        runner._log_message(f"[VideoTimestamps] RoundingMethod: {rounding_str}")

        # Thread-safe cache write
        with _vfr_cache_lock:
            _vfr_cache[cache_key] = vts
        return vts

    except ImportError:
        runner._log_message(
            "[VideoTimestamps] WARNING: VideoTimestamps not installed. Install with: pip install VideoTimestamps"
        )
        return None
    except Exception as e:
        runner._log_message(
            f"[VideoTimestamps] WARNING: Failed to create timestamps handler: {e}"
        )
        return None

File: crates/vsg_core/src/subtitles/frame_utils/timing.rs
//! Frame/time conversion functions.
//!
//! Pure functions for converting between frame numbers and timestamps.
//! All functions are deterministic and have no side effects.
//!
//! # Timing Modes
//!
//! - **Floor**: Frame START - stable, deterministic, preferred for sync math
//! - **Middle**: Middle of frame window - balanced approach
//! - **Aegisub**: Matches Aegisub's algorithm - ceil to centisecond

/// Small epsilon for floating-point comparisons.
const EPSILON: f64 = 1e-6;

// ============================================================================
// MODE 0: FRAME START (Floor-based, Deterministic)
// ============================================================================

/// Convert timestamp to frame number using FLOOR with epsilon protection.
///
/// This gives the frame that is currently displaying at the given time.
/// This is the preferred method for sync math because:
/// - Deterministic (no rounding ambiguity at boundaries)
/// - Stable under floating point drift
/// - Maps to actual frame boundaries (frame N starts at N * frame_duration)
///
/// # Arguments
/// * `time_ms` - Timestamp in milliseconds
/// * `fps` - Frame rate (e.g., 23.976)
///
/// # Returns
/// Frame number (which frame is displaying at this time)
///
/// # Examples
/// ```
/// use vsg_core::subtitles::frame_utils::timing::time_to_frame_floor;
///
/// // At 23.976 fps (frame_duration = 41.708ms):
/// assert_eq!(time_to_frame_floor(0.0, 23.976), 0);
/// assert_eq!(time_to_frame_floor(41.707, 23.976), 0);  // Still in frame 0
/// assert_eq!(time_to_frame_floor(41.709, 23.976), 1);  // Frame 1 starts
/// assert_eq!(time_to_frame_floor(1001.0, 23.976), 24); // Frame 24
/// ```
pub fn time_to_frame_floor(time_ms: f64, fps: f64) -> i32 {
    let frame_duration_ms = 1000.0 / fps;
    // Add small epsilon to protect against FP errors where time_ms is slightly under frame boundary
    ((time_ms + EPSILON) / frame_duration_ms) as i32
}

/// Convert frame number to its START timestamp (exact, no rounding).
///
/// This is the preferred method for sync math because:
/// - Frame N starts at exactly N * frame_duration
/// - No rounding (exact calculation)
/// - Guarantees frame-aligned timing
///
/// # Arguments
/// * `frame_num` - Frame number
/// * `fps` - Frame rate (e.g., 23.976)
///
/// # Returns
/// Timestamp in milliseconds (frame START time, as float for precision)
///
/// # Examples
/// ```
/// use vsg_core::subtitles::frame_utils::timing::frame_to_time_floor;
///
/// // At 23.976 fps (frame_duration = 41.708ms):
/// assert!((frame_to_time_floor(0, 23.976) - 0.0).abs() < 0.001);
/// assert!((frame_to_time_floor(1, 23.976) - 41.708).abs() < 0.01);
/// assert!((frame_to_time_floor(24, 23.976) - 1001.0).abs() < 0.1);
/// ```
pub fn frame_to_time_floor(frame_num: i32, fps: f64) -> f64 {
    let frame_duration_ms = 1000.0 / fps;
    frame_num as f64 * frame_duration_ms
}

// ============================================================================
// MODE 1: MIDDLE OF FRAME
// ============================================================================

/// Convert timestamp to frame number, accounting for +0.5 offset.
///
/// Uses the middle of the frame window for rounding decisions.
///
/// # Arguments
/// * `time_ms` - Timestamp in milliseconds
/// * `fps` - Frame rate (e.g., 23.976)
///
/// # Returns
/// Frame number
pub fn time_to_frame_middle(time_ms: f64, fps: f64) -> i32 {
    let frame_duration_ms = 1000.0 / fps;
    (time_ms / frame_duration_ms - 0.5).round() as i32
}

/// Convert frame to timestamp targeting middle of frame window.
///
/// Targets the middle of the frame's display window with +0.5 offset.
///
/// # Example at 23.976 fps
/// - Frame 24 displays from 1001.001ms to 1042.709ms
/// - Calculation: 24.5 × 41.708 = 1022ms
/// - After centisecond rounding: 1020ms (safely in frame 24)
///
/// # Arguments
/// * `frame_num` - Frame number
/// * `fps` - Frame rate (e.g., 23.976)
///
/// # Returns
/// Timestamp in milliseconds (rounded to integer)
pub fn frame_to_time_middle(frame_num: i32, fps: f64) -> i32 {
    let frame_duration_ms = 1000.0 / fps;
    ((frame_num as f64 + 0.5) * frame_duration_ms).round() as i32
}

// ============================================================================
// MODE 2: AEGISUB-STYLE (Ceil to Centisecond)
// ============================================================================

/// Convert timestamp to frame using floor division.
///
/// Matches Aegisub's time-to-frame conversion.
///
/// # Arguments
/// * `time_ms` - Timestamp in milliseconds
/// * `fps` - Frame rate
///
/// # Returns
/// Frame number
pub fn time_to_frame_aegisub(time_ms: f64, fps: f64) -> i32 {
    let frame_duration_ms = 1000.0 / fps;
    (time_ms / frame_duration_ms) as i32
}

/// Convert frame to timestamp using Aegisub's algorithm.
///
/// Matches Aegisub's algorithm: Calculate exact frame start, then round UP
/// to the next centisecond to ensure timestamp falls within the frame.
///
/// # Example at 23.976 fps
/// - Frame 24 starts at 1001.001ms
/// - Exact calculation: 24 × 41.708 = 1001.001ms
/// - Round UP to next centisecond: ceil(1001.001 / 10) × 10 = 1010ms
/// - Result: 1010ms (safely in frame 24: 1001-1043ms)
///
/// # Arguments
/// * `frame_num` - Frame number
/// * `fps` - Frame rate
///
/// # Returns
/// Timestamp in milliseconds (ceiled to centisecond)
pub fn frame_to_time_aegisub(frame_num: i32, fps: f64) -> i32 {
    let frame_duration_ms = 1000.0 / fps;
    let exact_time_ms = frame_num as f64 * frame_duration_ms;

    // Round UP to next centisecond (ASS format precision)
    // This ensures the timestamp is guaranteed to fall within the frame
    let centiseconds = (exact_time_ms / 10.0).ceil();
    (centiseconds * 10.0) as i32
}

// ============================================================================
// UTILITY FUNCTIONS
// ============================================================================

/// Calculate frame duration in milliseconds.
///
/// # Arguments
/// * `fps` - Frame rate
///
/// # Returns
/// Frame duration in milliseconds
#[inline]
pub fn frame_duration_ms(fps: f64) -> f64 {
    1000.0 / fps
}

/// Convert FPS to fraction for NTSC rates.
///
/// NTSC standards use fractional rates (N×1000/1001) to avoid color/audio drift.
///
/// # Arguments
/// * `fps` - Frame rate
///
/// # Returns
/// (numerator, denominator) tuple
pub fn fps_to_fraction(fps: f64) -> (u32, u32) {
    // Common NTSC framerates
    if (fps - 23.976).abs() < 0.01 {
        (24000, 1001) // 23.976fps - NTSC film
    } else if (fps - 29.97).abs() < 0.01 {
        (30000, 1001) // 29.97fps - NTSC video
    } else if (fps - 59.94).abs() < 0.01 {
        (60000, 1001) // 59.94fps - NTSC high fps
    } else if (fps - 24.0).abs() < 0.01 {
        (24, 1)
    } else if (fps - 25.0).abs() < 0.01 {
        (25, 1) // PAL
    } else if (fps - 30.0).abs() < 0.01 {
        (30, 1)
    } else if (fps - 50.0).abs() < 0.01 {
        (50, 1)
    } else if (fps - 60.0).abs() < 0.01 {
        (60, 1)
    } else {
        // Generic: multiply by 1000 and use that
        ((fps * 1000.0).round() as u32, 1000)
    }
}

/// Parse FPS fraction string (e.g., "24000/1001") to float.
///
/// # Arguments
/// * `s` - Fraction string like "24000/1001" or plain number like "25"
///
/// # Returns
/// FPS as float, or None if parsing fails
pub fn parse_fps_fraction(s: &str) -> Option<f64> {
    if s.contains('/') {
        let parts: Vec<&str> = s.split('/').collect();
        if parts.len() == 2 {
            let num: f64 = parts[0].trim().parse().ok()?;
            let denom: f64 = parts[1].trim().parse().ok()?;
            if denom != 0.0 {
                return Some(num / denom);
            }
        }
        None
    } else {
        s.trim().parse().ok()
    }
}

/// Generate candidate frame offsets centered on a correlation value.
///
/// # Arguments
/// * `correlation_frames` - Audio correlation converted to frames (can be fractional)
/// * `search_range_frames` - How many frames on each side to search
///
/// # Returns
/// Sorted list of integer frame offsets to test
pub fn generate_frame_candidates(correlation_frames: f64, search_range_frames: i32) -> Vec<i32> {
    let mut candidates = std::collections::HashSet::new();

    // Round correlation to nearest frame
    let base_frame = correlation_frames.round() as i32;

    // Always include zero (in case correlation is just wrong)
    candidates.insert(0);

    // Search window around correlation
    for delta in -search_range_frames..=search_range_frames {
        candidates.insert(base_frame + delta);
    }

    let mut result: Vec<i32> = candidates.into_iter().collect();
    result.sort();
    result
}

/// Select checkpoint times distributed across video duration.
///
/// Uses percentage-based positions, avoiding very start/end of video.
///
/// # Arguments
/// * `duration_ms` - Video duration in milliseconds
/// * `num_checkpoints` - Number of checkpoints to generate
///
/// # Returns
/// List of checkpoint times in milliseconds
pub fn select_checkpoint_times(duration_ms: f64, num_checkpoints: usize) -> Vec<f64> {
    // Use percentage-based positions (avoiding very start/end)
    let positions = [15, 30, 50, 70, 85];

    positions
        .iter()
        .take(num_checkpoints)
        .map(|&pos| duration_ms * pos as f64 / 100.0)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_time_to_frame_floor() {
        let fps = 23.976;

        assert_eq!(time_to_frame_floor(0.0, fps), 0);
        assert_eq!(time_to_frame_floor(41.0, fps), 0); // Still in frame 0
        assert_eq!(time_to_frame_floor(42.0, fps), 1); // Frame 1
        assert_eq!(time_to_frame_floor(1001.0, fps), 24); // Frame 24
    }

    #[test]
    fn test_frame_to_time_floor() {
        let fps = 23.976;

        assert!((frame_to_time_floor(0, fps) - 0.0).abs() < 0.001);
        assert!((frame_to_time_floor(24, fps) - 1001.0).abs() < 0.1);
    }

    #[test]
    fn test_aegisub_style() {
        let fps = 23.976;

        // Frame 24 should give 1010ms (ceiled to centisecond)
        let time = frame_to_time_aegisub(24, fps);
        assert_eq!(time, 1010);
    }

    #[test]
    fn test_fps_to_fraction() {
        assert_eq!(fps_to_fraction(23.976), (24000, 1001));
        assert_eq!(fps_to_fraction(29.97), (30000, 1001));
        assert_eq!(fps_to_fraction(25.0), (25, 1));
    }

    #[test]
    fn test_parse_fps_fraction() {
        assert!((parse_fps_fraction("24000/1001").unwrap() - 23.976).abs() < 0.001);
        assert!((parse_fps_fraction("25").unwrap() - 25.0).abs() < 0.001);
        assert!(parse_fps_fraction("invalid").is_none());
    }

    #[test]
    fn test_generate_frame_candidates() {
        // Correlation at 2.3 frames with range 3
        let candidates = generate_frame_candidates(2.3, 3);

        // Should include 0 (always) and -1 to 5 (2 ± 3)
        assert!(candidates.contains(&0));
        assert!(candidates.contains(&2));
        assert!(candidates.contains(&-1));
        assert!(candidates.contains(&5));
    }

    #[test]
    fn test_select_checkpoint_times() {
        let duration = 1000000.0; // 1000 seconds
        let checkpoints = select_checkpoint_times(duration, 5);

        assert_eq!(checkpoints.len(), 5);
        assert!((checkpoints[0] - 150000.0).abs() < 1.0); // 15%
        assert!((checkpoints[2] - 500000.0).abs() < 1.0); // 50%
    }

    #[test]
    fn test_roundtrip_floor() {
        let fps = 23.976;

        for frame in 0..100 {
            let time = frame_to_time_floor(frame, fps);
            let recovered = time_to_frame_floor(time, fps);
            assert_eq!(frame, recovered, "Roundtrip failed for frame {}", frame);
        }
    }
}
